internal/workflow/preprocessing.go

```diff
@@ -36,94 +55,225 @@ func (w *PreprocessingWorkflow) Execute(
+	}
+
+	if validateMetadata.Failures != nil {
+		validateMetadataEvent.Complete(
+			ctx,
+			enums.EventOutcomeInvalid,
+			"Content error: metadata validation has failed: %s",
+			strings.Join(validateMetadata.Failures, "\n"),
+		)
+	} else {
+		validateMetadataEvent.Complete(ctx, enums.EventOutcomeValid, "Metadata validation successful")
+	}
+	result.addEvent(validateMetadataEvent)
+
+	// Stop here if the SIP content isn't valid.
+	if !validateStructureEvent.IsValid() ||
+		!validateFileFormatsEvent.IsValid() ||
+		!validateMetadataEvent.IsValid() {
+		return contentError(&result)
 	}
 
 	// Re-structure SIP.
+	restructureSIPEvent := newEvent(ctx, "Restructure SIP")
 	var transformSIP activities.TransformSIPResult
 	e = temporalsdk_workflow.ExecuteActivity(
 		withLocalActOpts(ctx),
 		activities.TransformSIPName,
 		&activities.TransformSIPParams{SIP: identifySIP.SIP},
 	).Get(ctx, &transformSIP)
 	if e != nil {
-		return nil, e
+		result.addEvent(restructureSIPEvent.Complete(
+			ctx,
+			enums.EventOutcomeFailure,
+			"System error: restructuring has failed",
+		))
+		return systemError(logger, "Restructure SIP", &result, e)
 	}
 
+	result.addEvent(restructureSIPEvent.Complete(
+		ctx,
+		enums.EventOutcomeSuccess,
+		"SIP has been restructured",
+	))
+
 	// Bag the SIP for Enduro processing.
+	createBagEvent := newEvent(ctx, "Bag SIP")
 	var createBag bagit.CreateBagActivityResult
 	e = temporalsdk_workflow.ExecuteActivity(
 		withLocalActOpts(ctx),
 		bagit.CreateBagActivityName,
 		&bagit.CreateBagActivityParams{SourcePath: localPath},
 	).Get(ctx, &createBag)
 	if e != nil {
-		return nil, e
+		result.addEvent(createBagEvent.Complete(
+			ctx,
+			enums.EventOutcomeFailure,
+			"System error: bagging has failed",
+		))
+		return systemError(logger, "Create bag", &result, e)
 	}
 
+	result.addEvent(createBagEvent.Complete(
+		ctx,
+		enums.EventOutcomeSuccess,
+		"SIP has been bagged",
+	))
+
 	// TODO: validate checksums located in the XML metadata file
 	// against the checksums generated on Bag creation.
 
-	return &PreprocessingWorkflowResult{RelativePath: params.RelativePath}, e
+	return &result, nil
 }
 
 func withLocalActOpts(ctx temporalsdk_workflow.Context) temporalsdk_workflow.Context {
 	return temporalsdk_workflow.WithActivityOptions(ctx, temporalsdk_workflow.ActivityOptions{
-		ScheduleToCloseTimeout: 5 * time.Minute,
+		ScheduleToCloseTimeout: time.Minute,
 		RetryPolicy: &temporalsdk_temporal.RetryPolicy{
 			MaximumAttempts: 1,
 		},
 	})
 }
+
+//nolint:unparam
+func contentError(result *PreprocessingWorkflowResult) (*PreprocessingWorkflowResult, error) {
+	result.Outcome = OutcomeContentError
+	return result, nil
+}
+
+//nolint:unparam
+func systemError(
+	logger temporalsdk_log.Logger,
+	activityName string,
+	result *PreprocessingWorkflowResult,
+	err error,
+) (*PreprocessingWorkflowResult, error) {
+	logger.Error(activityName, "system error", err.Error())
+	result.Outcome = OutcomeSystemError
+	return result, nil
+}
```

Comment:
Not a big deal either, feel free to ignore.

These two functions feel like a bit of magic to me, with `nolint:unparam` and returning things that should be needed just to be able to use them directly from a return call. I get why you create them, but I think the first one is not needed (only used once), and the second one I'd have it without return, just logging and updating the result, then calling return after the function call.

Reply:
I'll get rid of the unused "error" return, which removes the need for the `nolint:unparam` directive.

Even though `contentError()` is only called once,  I think it's useful as documentation in code of what the expected returns from a content error are - a developer can easily look at the `contentError()` and `systemError()` function definitions to understand what the workflow should return in each case.